src/character.cpp

```diff
@@ -8884,19 +8888,103 @@ bool Character::in_sleep_state() const
+    visit_items( [this, &key, &flag]( item * it, item * ) {
+        if( it->has_flag( flag ) ) {
+            inv_search_caches[key].items.push_back( it );
+        }
+        return VisitResponse::NEXT;
+    } );
+
+    return inv_search_caches[key].items;
+}
+
+std::vector<item *> &Character::all_items_with( const std::string &key,
+        bool( item::*filter_func )() const ) const
+{
+    auto iter = inv_search_caches.find( key );
+    if( iter != inv_search_caches.end() ) {
+        // If the cache already exists, use it.
+        return iter->second.items;
+    }
+    // Otherwise, add a new cache and populate with all appropriate items in the inventory. Empty sets are still created.
+    inv_search_caches[key].filter_func = filter_func;
+    visit_items( [this, &key, &filter_func]( item * it, item * ) {
+        if( ( it->*filter_func )() ) {
+            inv_search_caches[key].items.push_back( it );
+        }
+        return VisitResponse::NEXT;
+    } );
+
+    return inv_search_caches[key].items;
+}
+
+std::vector<item *> &Character::all_items_with( const std::string &key, const flag_id &flag,
+        bool( item::*filter_func )() const ) const
+{
+    auto iter = inv_search_caches.find( key );
+    if( iter != inv_search_caches.end() ) {
+        // If the cache already exists, use it.
+        return iter->second.items;
+    }
+    // Otherwise, add a new cache and populate with all appropriate items in the inventory. Empty sets are still created.
+    inv_search_caches[key].flag = flag;
+    inv_search_caches[key].filter_func = filter_func;
+    visit_items( [this, &key, &flag, &filter_func]( item * it, item * ) {
+        if( it->has_flag( flag ) && ( it->*filter_func )() ) {
+            inv_search_caches[key].items.push_back( it );
         }
-        return it.has_flag( flag );
+        return VisitResponse::NEXT;
     } );
+
+    return inv_search_caches[key].items;
+}
+
+void Character::add_to_inv_search_caches( item &it ) const
+{
+    for( auto &cache : inv_search_caches ) {
+        if( cache.second.flag.is_valid() && !it.has_flag( cache.second.flag ) ) {
+            continue;
+        }
+        if( cache.second.filter_func && !( it.*cache.second.filter_func )() ) {
+            continue;
+        }
+        cache.second.items.push_back( &it );
+    }
 }
 
-std::vector<const item *> Character::all_items_with_flag( const flag_id &flag ) const
+void Character::remove_from_inv_search_caches( item &it ) const
 {
-    return items_with( [&flag]( const item & it ) {
-        return it.has_flag( flag );
+    it.visit_items( [this]( item * i, item * ) {
+        for( auto &cache : inv_search_caches ) {
+            if( cache.second.flag.is_valid() && !i->has_flag( cache.second.flag ) ) {
+                continue;
+            }
+            if( cache.second.filter_func && !( i->*cache.second.filter_func )() ) {
+                continue;
+            }
+            inv_search_caches.erase( cache.first );
+        }
```

Comment:
It crashes here. Example fix:

```suggestion
        for( auto it = inv_search_caches.begin(); it != inv_search_caches.end(); ) {
            if( ( !it->second.flag.is_valid() || i->has_flag( it->second.flag ) ) &&
                ( it->second.filter_func == nullptr || ( i->*it->second.filter_func )() ) ) {
                it = inv_search_caches.erase( it );
            } else {
                ++it;
            }
        }
```